### ISSUE TITLE
Blueprints names and path problems

### ISSUE THREAD

**Issue by Romarain** [getgrav/grav#2708]
When placing my blueprints in `user/themes/my_theme/blueprints` : 
1. Using a "modular.yaml" or an "external.yaml" blueprint (as in the original `system/blueprints` folder) won't work.
2. The `context: blueprints://` doesn't work : wether we use `context: blueprints://pages` or `context: blueprints://santa_claus`, nothing work better or worse.
3. It appears that we can't separate blueprints into several subfolders, for example in order to make the `modular.yaml` to work. I tried to place it in a `/modular` sub-folder, without success.
4. Worse : naming it "default.yaml" after having placed it in the `/modular` subfolder, makes the page **to crash** ! (error 500).

To make a modular blueprint to work for all type of modular page, I figured that using the `user/blueprints/pages/modular` folder instead, was working ! But I have to rename the file "default.yaml" for it to be used.

So, clearly, two things are highlighted in those tests : 
1. "default.yaml" is the only name accepted for rewritting the original blueprints _(we can still use any other name related to our templates, fortunately !)_
2. The `user/themes/my_theme/blueprints` doesn't respect the original blueprint directory structure _(while `user/blueprints/pages/modular` do !)_.
3. To create a generic modular blueprint, we have to place it in `user/blueprints/pages/modular`, wich kills any possibility to link it to a theme.

**Comment by rhukster** [comment]
I'm not really following this at all.  the user 'blueprints' only overrides the blueprints/pages system folder. so `mytheme/blueprints/default.yaml` overrides the `system/blueprints/pages/default.yaml` file, just as `mytheme/blueprints/modular.yaml` overrides the `system/blueprints/pages/modular.yaml` file.  

In general overriding the default blueprint is not advisable as default template blueprint is usually inherited by all page types.  A betters solution is to change **all** your pages to extend some other default template/blueprint, and then override that.

**Comment by Romarain** [comment]
> the user 'blueprints' only overrides the blueprints/pages system folder.
Ok
> so mytheme/blueprints/default.yaml overrides the system/blueprints/pages/default.yaml file
Ok
> just as mytheme/blueprints/modular.yaml overrides the system/blueprints/pages/modular.yaml file.
It doesn't seem to be.

Anyway, the problems of path I mentioned are real. I feel a bit disturbed to repeat the information I already written, but the way the two paths (meaning `user/themes/my_theme/blueprints` and `user/blueprints/`) are handled is problematic.

**Comment by OleVik** [comment]
I'll try an interpretation, to answer the first four points:

1. Are you referencing these in your blueprints? We can't replicate the issue without knowing whether you are extending, importing, overriding, or otherwise manipulating these files.
2. Note that these are streams, which resolve to physical locations within Grav. `context: blueprints://` is not apt in the case you describe, did you mean `user://themes/my_theme/blueprints` or even `theme://blueprints`?
3. You can, depending on the situation. As explained in the very recently proposed [755 on Docs](https://github.com/getgrav/grav-learn/pull/755), this:

```yaml
extends@:
  type: default
  context: blueprints://`
```

Will resolve to `/user/plugins/admin/blueprints/default.yaml` by default. You can use slashes in `type` though, and use something like:

```yaml
extends@:
  type: 'modular/default'
  context: 'user://themes/my_theme/blueprints'
```

4. That sounds unlikely and not related, but cannot be understood without a stack trace and minimal working example that can be replicated.

**Comment by Romarain** [comment]
@OleVik Im' sorry but I don't understand you. I don't want to be rude, but your answer is causing me a maximum of effort to understand anything and be able to answer something coherent with the discussion.

> Are you referencing these in your blueprints?

Referencing ? What do you mean ? We are supposed to be allowed to recreate the same files (same name and extension) in a personal blueprints folder. Period. To override or extend the `default.yaml`, we just have to create a `default.yaml` and fill it correctly. Same for `modular.yaml`.

> We can't replicate the issue without knowing whether you are extending, importing, overriding, or otherwise manipulating these files.

You don't have to. I'm exposing a problem with the `context:` option, wich indicates that I'm using it to import files, maybe, but even if I was using it in the first line of the default.yaml _(wich was sometimes proposed by people who helped me to debug my blueprints)_, it will just be for overriding that default.yaml, nothing more. **My answer might pass for a non-answer, but it means that despite any specific purpose, all the attempt of using the `context:` options failed.**

To be clear (again), it just doesn't do less or more than if I don't use the option, or if I use "sant_claus" as a value.

> context: blueprints:// is not apt in the case you describe, did you mean user://themes/my_theme/blueprints or even theme://blueprints?

No, I meant what I meant, and that is shown in the documentation. I don't invent things ;)
What you propose here is totally new to my eyes.

> You can, depending on the situation. As explained in the very recently proposed 755 on Docs, this:

> Will resolve to /user/plugins/admin/blueprints/default.yaml by default

So now, `context: blueprints://` isn't so useless ? It has a sense.

Can we have a documentation on those blueprints streams/whatever you call it ?
I would like to understand how and where to place my blueprints, and how streams have to be used.

**For now, things are not working properly, has I described it in this ticket : placing files in a similar sub-location doesn't work (I won't repeat what I said in my first bullet list, it would be useless), and the `context:` option has no effect.**

So I require more explanation on how override the original `/system/blueprints/` files, how replicate the folders structure (ex : `/system/blueprints/pages/`), and how to use the `context:` option (may necessitate a full documentation).

**Comment by Romarain** [comment]
> That sounds unlikely and not related, but cannot be understood without a stack trace and minimal working example that can be replicated.

The working example consist, as I said, to try to override the modular.yaml by creating it in the `user/themes/my_theme/blueprints`, as I was talking about since the begining. Then, to move it in a /modular sub-folder, wich no more success, and to rename it "default.yaml".

Now, I'm not on my dev-PC, but here's a map to replicate all the problems I talked about.

A1. In `user/themes/my_theme/blueprints`, create a `default.yaml`, a `test.yaml`, and a `modular.yaml`.
A2. Also, create a `/modular` sub-folder.
A3. In `default.yaml`, place a (I do it from memory here) : 
```
extends@: default

form:
    fields:
        tabs:
            content:
            type: tab
            replace@: true
            fields:
                import@: 'test.yaml'
                context: 'blueprints://'
```
A4. In test.yaml, do what you want, it has no utility.
A5. In modular.yaml, do what you want.

Now, has a test : 

B1 : In `default.yaml`, modify the `context:` option, like `context: 'blueprints://santa_claus'`. Observe how it has no effect.
B2 : Observe how `modular.yaml` doesn't override anything in a modular page's admin edition. Fields that you might have created aren't showing.
B3 : In that `modular.yaml`, add a context in the extend@ : 
```
extends@: 
    type: default
    context: blueprints://
```
B4 : Play with that context, has you want. Move the file (modular.yaml) in the `/modular` sub-folder and adapt the context option with (for instance) `context: blueprints://modular`. Observe how it has no effect.

I don't even talk about the `type:` option, that I also don't understand. It seems that using something else than "default" as a value makes the blueprint to stop working, randomly. **I have no information about how to use that option**. Seems logical to think that it may be related to the yaml name, but at the same time, seems not that necessary (a "type" isn't a name !).

**Comment by OleVik** [comment]
Yes, as I say, an addition to that documentation is proposed, and Streams are already fleshed out in the multisite docs. As far as I know, /blueprints are auto-loaded in themes, but any other locations or in plugins would need to be manually loaded - or resolved with `context`.

I agree that debugging blueprints is difficult, and developing advanced usage of them is not for the faint of heart. As said above, you need to consider `type` as a reference to a filename, and `context` something that resolves into a folder, when using `extends@`. I'd like to document more things about blueprints more fully, but I too struggle with them - so we're left in the situation that the only way to improve the precision, clarity, consistency, and coverage of the docs on this subject is to find out what works and document that.
